### ISSUE TITLE
Litedown: Support mailto: in markdown links

### ISSUE THREAD

**Issue by tobscure** [s9e/TextFormatter#109]
Currently the following:

```
[email](mailto:toby@flarum.org)
```

is rendered as:

> \[email](mailto:[toby@flarum.org](mailto:toby@flarum.org))

instead of:

> [email](mailto:toby@flarum.org)

It would be nice if Litedown could support `mailto:` links like other common Markdown implementations (including GitHub-Flavoured Markdown and CommonMark)

**Comment by JoshyPHP** [comment]
It would be difficult to implement properly at the moment. The main reason is that the library doesn't have a URI type/filter. It has [a `#url` filter](https://s9etextformatter.readthedocs.io/Filters/Built-in_filters/) that handles URLs but not all URIs. Litedown links translate into the same `URL` tag as the `[url]` BBCode, and the `#url` filter rejects URLs with no "authority" part. (the part that starts with `//`) Or rather, the filter doesn't identify the authority part of a mailto: URI.

I'm considering various options such as making the names of the tags used by the Litedown plugin configurable, or adding a URI filter. I'd have to more or less duplicate the [URL configurator](https://s9etextformatter.readthedocs.io/Filters/URL_features/Schemes/) for the latter, which seems redundant. I don't know what's the best long-term plan and I'm open to feedback.

In the meantime, you can add basic support for mailto: links in userland. The simplest way would be to use the Preg plugin with something like this:

```php
$configurator->Preg->match(
	'/\\[(.*?)\\]\\(mailto:(?<email>[^@]+@[-.\\w]+)\\)/',
	'EMAIL'
);
```

It wouldn't support queries (e.g. prefilled subject, cc, etc...) or link titles. The alternative would be to add a tag filter to post-process the `URL` tag and create an `EMAIL` tag in userland where applicable.

```php
$configurator = new s9e\TextFormatter\Configurator;
$configurator->Autoemail;
$configurator->Litedown;

$configurator->tags['URL']->filterChain->prepend('filterUrlMailto')
	->resetParameters()
	->addParameterByName('tag')
	->addParameterByName('parser');

function filterUrlMailto($tag, $parser)
{
	$endTag = $tag->getEndTag();
	if (!$endTag || !preg_match('/^mailto:([^@]+@[-.\\w]+)/', $tag->getAttribute('url'), $m))
	{
		return;
	}

	$parser->addTagPair('EMAIL', $tag->getPos(), $tag->getLen(), $endTag->getPos(), $endTag->getLen())
	       ->setAttribute('email', $m[1]);
}

extract($configurator->finalize());

$text = '[..](mailto:example@example.org)';
$xml  = $parser->parse($text);
$html = $renderer->render($xml);

die("$html\n");
```
```html
<p><a href="mailto:example@example.org">..</a></p>
```

You could also remove the default `#url` filter from the `URL` tag and add your own:

```php
$configurator = new s9e\TextFormatter\Configurator;
$configurator->Litedown;

$filterChain = $configurator->tags['URL']->attributes['url']->filterChain;
$filterChain->remove('#url');
$filterChain->append('myUrlFilter')
	->resetParameters()
	->addParameterByName('attrValue')
	->addParameterByName('urlConfig')
	->addParameterByName('logger');

function myUrlFilter($attrValue, $urlConfig, $logger)
{
	if (preg_match('(^mailto:)', $attrValue))
	{
		return $attrValue;
	}

	return s9e\TextFormatter\Parser\AttributesFilters\UrlFilter::filter($attrValue, $urlConfig, $logger);
}

extract($configurator->finalize());

$text = '[..](mailto:example@example.org)';
$xml  = $parser->parse($text);
$html = $renderer->render($xml);

die("$html\n");
```

It's also possible to redefine the `#url` filter globally (not just for a given tag) but that's more complicated and probably not that useful.

**TL;DR**: there's currently no support for URIs and I don't know when there will be. In the meantime there are many ways to implement a workaround in userland, each with their own tradeoffs.

**Comment by JoshyPHP** [comment]
Sorry for not replying to your last comment sooner, I either missed the notification or read it and forgot about it. 😳 I'm setting this issue to be addressed in the 2.0 milestone.

Currently, #url cannot parse all URIs. It only groks relative URLs or URLs with an *authority* component. My intention was to add a generic #uri filter if the need arose. However, as the living standards have evolved the meaning of "URL" to effectively mean any URI that isn't specifically a URN, it makes sense for the #url filter to accept those.

The primary goal for the #url filter is to control which schemes are allowed and to sanitize the value so that it can be safely used in CSS and JavaScript. It also encodes IDN hosts and controls which hosts are allowed. That last part requires to be able to identify the host in a given URL. Strict validation isn't one of the goal.

My intention is to update the #url filter in 2.0 to allow what HTML 5.2 describes as a URL. I need to be careful and not accidentally allow things like `http:example.org` which could be used to bypass the host checker.

**Comment by JoshyPHP** [comment]
As of 2.0, you should be able to allow mailto: links by [allowing the mailto scheme](https://s9etextformatter.readthedocs.io/Filters/URL_features/Schemes/).

```php
$configurator->urlConfig->allowScheme('mailto');
```
